File: limited_time_mfpt/code/run_random_walk.R
r_run_random_walk <- function(P, start, numSteps, numRuns, movement, filename)
{
    N <- nrow(P)
    walker <- matrix(data = NaN, nrow= numRuns, ncol = numSteps)
    for(run in 1:numRuns)
    {
        walker[run,1] <- start
        for(step in 2:numSteps)
        {
            loc <- walker[run, step - 1]
            p <- P[loc,]
            p <- cumsum(p)
            aux <- p - movement[run, step]
            walker[run, step] = which(aux >= 0)[1]
        }
    }
    
    write.csv(walker, file = filename)
    
    lt_mfpt <- matrix(data = NaN, nrow = numSteps, ncol = N)
    fraction_arrival <- matrix(data = NaN, nrow = numSteps, ncol = N)
    numArr <- matrix(data = 0, nrow = numSteps, ncol = N)
    
    for(target in 1:N){
        # Addition of clause such that MFPT and FracArr and narr do not include start condition for the walker matrix
        if(target == start){
            #print("Before")
            twalker = walker
            walker[,1] = 0
        }
        for(time in 1:numSteps)
        {
            walker_limited <- matrix(walker[,1:time], ncol = time)
            arrived <- walker_limited[which(walker_limited == target, arr.ind = TRUE)[,1],]
            fraction_arrival[time, target] <- length(unique(which(walker_limited == target, 
                                                                  arr.ind = TRUE)[,1])) / numRuns
            
            if(length(arrived) > 0)
            {
                if(is.null(ncol(arrived)))
                {
                    lt_mfpt[time, target] <- min(which(arrived == target, arr.ind = TRUE))
                    if(time == 1 && target == start){
                        numArr[time, target] <- 1
                    }
                    else{
                        numArr[time, target] <- length(which(arrived == target, arr.ind = TRUE))
                    }
                }
                else
                {
                    lt_mfpt[time, target] <- mean(apply(arrived, 1, 
                                                        function(x) min(which(x == target, arr.ind = TRUE))), na.RM=T)
                    numArr[time, target] <- mean(apply(arrived, 1, 
                                                  function(x) length(which(x == target, arr.ind = TRUE))), na.RM=T)
                }
            }
        }
        # Add clause for mean return time to set it such that the mfpt and fracArr cannot include the initial step
        # This returns the walker matrix to the original
        if(target == start){
            #print("After")
            walker = twalker
        }
    }
    
    return(list(fraction_arrival=fraction_arrival, lt_mfpt= lt_mfpt , numArr = numArr))
}

File: limited_time_mfpt/code/Full/run_random_walk.m
function [fraction_arrival,lt_mfpt]= run_random_walk(P,source,target,numSteps,numRuns)

walker = nan(numRuns,numSteps);
for run=1:numRuns
    %disp(run)
    walker(run,1) = source;
    step=2;
    while (step<=numSteps) && (walker(run,step-1)~=target)
        p = P(walker(run,step-1),:);
        p = cumsum(p);
        movement = rand;
        aux = p - movement;
        walker(run,step) = find(aux>=0,1);
        step = step + 1;
    end
end

lt_mfpt = nan(numSteps,1);
fraction_arrival = nan(numSteps,1);
for time=1:numSteps
    walker_limited=walker(:,1:time);
    [rows,cols]=find(walker_limited==target);
    fraction_arrival(time) = length(rows)/numRuns;
    if ~isempty(cols)
        lt_mfpt(time) = mean(cols);
    end
end
%figure, imagesc(walker); xlabel('time'); ylabel('runs'); colorbar;
%figure, plot(lt_mfpt); xlabel('time'); ylabel('lt_mfpt');
%figure, plot(fraction_arrival); xlabel('time'); ylabel('fraction arrival');



    


